src/org/exist/source/URLSource.java

```diff
@@ -120,6 +126,26 @@ public Validity isValid(final Source other) {
 		return Validity.INVALID;
 	}
 
+	public Charset getEncoding() throws IOException {
+		if (connection == null) {
+			connection = url.openConnection();
+		}
+		final String contentType = connection.getContentType();
+		if (contentType != null) {
+			final Pattern pattern = PatternFactory.getInstance().getPattern("(?i)\\bcharset=\\s*\"?([^\\s;\"]*)");
```

Comment:
Using the PatternFactory cache here doesn't make much sense as the regex is static. Instead maybe just compile as a `private static final` member variable.